lib/Dialect/FIRRTL/Transforms/GrandCentralTaps.cpp

```diff
@@ -0,0 +1,653 @@
+#include "circt/Dialect/FIRRTL/FIRRTLOps.h"
+#include "circt/Dialect/FIRRTL/FIRRTLTypes.h"
+#include "circt/Dialect/FIRRTL/FIRRTLVisitors.h"
+#include "circt/Dialect/FIRRTL/Passes.h"
+#include "circt/Dialect/SV/SVOps.h"
+#include "mlir/IR/ImplicitLocOpBuilder.h"
+#include "llvm/ADT/STLExtras.h"
+#include "llvm/ADT/StringExtras.h"
+#include "llvm/Support/Allocator.h"
+#include "llvm/Support/Debug.h"
+#include "llvm/Support/FormatVariadic.h"
+
+#define DEBUG_TYPE "gct"
+
+using namespace circt;
+using namespace firrtl;
+using mlir::function_like_impl::getArgAttrDict;
+using mlir::function_like_impl::setAllArgAttrDicts;
+
+//===----------------------------------------------------------------------===//
+// Utilities
+//===----------------------------------------------------------------------===//
+
+/// An absolute instance path.
+using InstancePath = ArrayRef<InstanceOp>;
+
+template <typename T>
+static T &operator<<(T &os, const InstancePath &path) {
+  os << "$root";
+  for (auto inst : path)
+    os << "." << inst.name();
+  return os;
+}
+
+namespace {
+
+/// A port annotated with a data tap key or mem tap.
+struct AnnotatedPort {
+  unsigned portNum;
+  Annotation anno;
+};
+
+/// An extmodule that has annotated ports.
+struct AnnotatedExtModule {
+  FExtModuleOp extModule;
+  SmallVector<AnnotatedPort, 4> portAnnos;
+  // Module annotations without data tap stuff.
+  ArrayAttr filteredModuleAnnos;
+  /// Port annotations without data tap stuff.
+  SmallVector<AnnotationSet, 8> filteredPortAnnos;
+};
+
+/// A value annotated to be tapped.
+struct TappedValue {
+  Value value;
+  Annotation anno;
+};
+
+/// A data structure that tracks the instances of modules and can provide
+/// absolute paths to these instances.
+struct InstanceGraph {
+  /// The root circuit.
+  CircuitOp circuitOp;
+  /// The main module in the circuit.
+  Operation *mainModule;
+  /// A mapping from a module to all its instances in the design.
+  DenseMap<Operation *, SmallVector<InstanceOp, 1>> moduleInstances;
+
+  InstanceGraph(CircuitOp circuitOp);
+  ArrayRef<InstancePath> getAbsolutePaths(Operation *op);
+
+private:
+  /// An allocator for individual instance paths and entire path lists.
+  llvm::BumpPtrAllocator allocator;
+
+  /// Cached absolute instance paths.
+  DenseMap<Operation *, ArrayRef<InstancePath>> absolutePathsCache;
+
+  /// Append an instance to a path.
+  InstancePath appendInstance(InstancePath path, InstanceOp inst);
+};
+
+/// Necessary information to wire up a port with tapped data or memory location.
+struct PortWiring {
+  unsigned portNum;
+  ArrayRef<InstancePath> prefices;
+  SmallString<16> suffix;
+};
+
+} // namespace
+
+InstanceGraph::InstanceGraph(CircuitOp circuitOp) : circuitOp(circuitOp) {
+  mainModule = circuitOp.getMainModule();
+
+  // Gather all instances in the circuit.
+  circuitOp.walk([&](InstanceOp op) {
+    moduleInstances[op.getReferencedModule()].push_back(op);
+  });
```

Comment:
Nice.  Perhaps someday this can use the InstanceGraph.

Reply:
Yeah excellent idea! Switched over to `firrtl::InstanceGraph`, now that we have it. Code reuse :tada: